File: android/src/main/java/ZoomProcessors/LivenessCheckProcessor.java
// Demonstrates performing a Liveness Check.

package ZoomProcessors;

import android.content.Context;
import org.json.JSONObject;

import com.facetec.zoom.sdk.*;

public class LivenessCheckProcessor extends Processor implements ZoomFaceMapProcessor {
    ZoomFaceMapResultCallback zoomFaceMapResultCallback;
    ZoomSessionResult latestZoomSessionResult;
    private boolean _isSuccess = false;
    private final String licenseKey;

    public LivenessCheckProcessor(Context context, String licenseKey) {
        // Launch the ZoOm Session.
        this.licenseKey = licenseKey;
        ZoomSessionActivity.createAndLaunchZoomSession(context, this);
    }

    public boolean isSuccess() {
        return _isSuccess;
    }

    // Required function that handles calling ZoOm Server to get result and decides how to continue.
    public void processZoomSessionResultWhileZoomWaits(final ZoomSessionResult zoomSessionResult, final ZoomFaceMapResultCallback zoomFaceMapResultCallback) {
        this.latestZoomSessionResult = zoomSessionResult;
        this.zoomFaceMapResultCallback = zoomFaceMapResultCallback;

        // Cancel last request in flight.  This handles case where processing is is taking place but cancellation or Context Switch occurs.
        // Our handling here ends the latest in flight request and simply re-does the normal logic, which will cancel out.
        NetworkingHelpers.cancelPendingRequests();

        // cancellation, timeout, etc.
        if (zoomSessionResult.getStatus() != ZoomSessionStatus.SESSION_COMPLETED_SUCCESSFULLY) {
            zoomFaceMapResultCallback.cancel();
            this.zoomFaceMapResultCallback = null;
            return;
        }

        // Create and parse request to ZoOm Server.
        NetworkingHelpers.getLivenessCheckResponseFromZoomServer(licenseKey, zoomSessionResult, this.zoomFaceMapResultCallback, new FaceTecManagedAPICallback() {
            @Override
            public void onResponse(JSONObject responseJSON) {
                UXNextStep nextStep = ServerResultHelpers.getLivenessNextStep(responseJSON);

                if (nextStep == UXNextStep.Succeed) {
                    _isSuccess = true;
                    ZoomCustomization.overrideResultScreenSuccessMessage = "Liveness\nConfirmed";
                    zoomFaceMapResultCallback.succeed();
                }
                else if (nextStep == UXNextStep.Retry) {
                    zoomFaceMapResultCallback.retry();
                }
                else {
                    zoomFaceMapResultCallback.cancel();
                }
            }
        });

    }
}

File: ios/LivenessCheckProcessor.swift
// Demonstrates performing a Liveness Check.

import UIKit
import Foundation
import FaceTecSDK

class LivenessCheckProcessor: NSObject, URLSessionDelegate, FaceTecFaceScanProcessorDelegate {
    var licenseKey: String!
    var delegate: ProcessingDelegate
    var latestZoomSessionResult: FaceTecSessionResult?
    var isSuccess: Bool

    init(sessionToken: String, delegate: ProcessingDelegate, fromVC: UIViewController, options: Dictionary<String, Any>) {
        self.delegate = delegate
        self.isSuccess = false

        super.init()
        
        self.licenseKey = options["licenseKey"] as? String

        // Launch the ZoOm Session.
        let sessionVC = FaceTec.sdk.createSessionVC(faceScanProcessorDelegate: self, sessionToken: sessionToken)
        if (options["useOverlay"] as? Bool ?? false) {
            sessionVC.modalPresentationStyle = .overCurrentContext
        }
        
        fromVC.present(sessionVC, animated: true, completion: nil)
    }
    
    // Required function that handles calling ZoOm Server to get result and decides how to continue.
    func processSessionWhileFaceTecSDKWaits(sessionResult: FaceTecSessionResult, faceScanResultCallback: FaceTecFaceScanResultCallback) {
        self.latestZoomSessionResult = sessionResult
        // cancellation, timeout, etc.
        if sessionResult.status != .sessionCompletedSuccessfully {
            faceScanResultCallback.onFaceScanResultCancel();
            return
        }
        
        // Create and parse request to ZoOm Server.
        NetworkingHelpers.getLivenessCheckResponseFromZoomServer(
            urlSessionDelegate: self,
            licenseKey: self.licenseKey,
            zoomSessionResult: sessionResult,
            resultCallback: { nextStep in
                if nextStep == .Succeed {
                    // Dynamically set the success message.
//                    ZoomCustomization.setOverrideResultScreenSuccessMessage("Liveness\nConfirmed")
                    faceScanResultCallback.onFaceScanResultSucceed()
                    self.isSuccess = true
                }
                else if nextStep == .Retry {
                    faceScanResultCallback.onFaceScanResultRetry()
                }
                else {
                    faceScanResultCallback.onFaceScanResultCancel()
                }
            }
        )
    }
    
    // iOS way to get upload progress and update ZoOm UI.
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
//        let uploadProgress: Float = Float(totalBytesSent) / Float(totalBytesExpectedToSend)
    }
    
    // The final callback ZoOm SDK calls when done with everything.
    func onFaceTecSDKCompletelyDone() {
        delegate.onProcessingComplete(isSuccess: isSuccess, facetecSessionResult: latestZoomSessionResult)
    }
}
